app.py

```diff
@@ -1,83 +1,56 @@
-    ),
-    (
-        r"/font-awesome/(.*)",
-        tornado.web.StaticFileHandler,
-        {"path": WebApp.PATH + "/static/font-awesome"},
-    ),
-    (
-        r"/js/(.*)",
-        tornado.web.StaticFileHandler,
-        {"path": WebApp.PATH + "/static/js"},
-    ),
-    (
-        r"/(favicon.ico|menu.html)",
-        tornado.web.StaticFileHandler,
-        {"path": WebApp.PATH + "/static/"},
-    ),
-]
-
-if __name__ == "__main__":
-    handlers = handlers.get_handlers()
-    handlers.extend(static_handlers)
-    app = WebApp(handlers)
-    http_server = tornado.httpserver.HTTPServer(app)
-    port = int(os.environ["PORT"])
-    num_processes = int(os.environ.get("NUM_PROCESSES", 1))
-    http_server.bind(port)
-    http_server.start(num_processes)
-    app.mongo = get_mongo()
-    app.redis = get_redis()
-    app.limit = int(os.environ.get("LIMIT", 10))
-    app.period = int(os.environ.get("PERIOD", 20))
-    app.videos = config.videos
-    try:
-        date = datetime.strptime(os.environ.get("GAME_DATE", ""), '%Y-%m-%d').date()
-        delta = (datetime.utcnow().date() - date).days
-    except ValueError:
-        delta = 0
-    app.days_delta = delta
-
-    while True:
-        logger.warning("Running app on port %d", port)
-        try:
-            tornado.ioloop.IOLoop.current().start()
-        except Exception as ex:
-            logger.exception(ex)
-            pass
+from common.session import get_mongo, get_redis
+
+from fastapi import FastAPI, Request, HTTPException
+
+from handlers import router
+
+app = FastAPI()
+app.state.mongo = get_mongo()
+app.state.redis = get_redis()
+app.state.limit = int(os.environ.get("LIMIT", 10))
+app.state.period = int(os.environ.get("PERIOD", 20))
+app.state.videos = config.videos
+app.state.current_timeframe = 0
+app.state.usage = defaultdict(int)
+app.state.api_key = config.api_key
+app.state.quotes = config.quotes
+app.state.js_version = uuid.uuid4().hex[:6]
+try:
+    date = datetime.strptime(os.environ.get("GAME_DATE", ""), '%Y-%m-%d').date()
+    delta = (datetime.utcnow().date() - date).days
+except ValueError:
+    delta = 0
+app.state.days_delta = timedelta(days=delta)
+app.mount("/static", StaticFiles(directory="static"), name="static")
+app.include_router(router)
+
+
+def request_is_limited(key: str):
+    now = int(time.time())
+    current = now - now % app.state.period
+    if app.state.current_timeframe != current:
+        app.state.current_timeframe = current
+        for ip, usage in app.state.usage.items():
+            if usage > app.state.limit * 0.75:
+                app.state.usage[ip] = usage // 2
+            else:
+                app.state.usage[ip] = 0
+    app.state.usage[key] += 1
+    return app.state.usage[key] > app.state.limit
+
+
+@app.middleware("http")
+async def is_limited(request: Request, call_next):
+    ip_address = request.client.host
+    if request_is_limited(key=ip_address):
+        raise HTTPException(status_code=429, detail="Limited!")
```

Comment:
right now this returns 500, I think it's due to the fact it's in the middlewhere and not the handler itself. Take a look at: https://stackoverflow.com/questions/61358669/raise-exception-in-python-fastapi-middleware

Reply:
032dc228c2ccc76bd24ee0ec5ef107262a8c0c9e

Reply:
![image](https://user-images.githubusercontent.com/11427786/165280500-2e947007-1f2a-4658-a412-e5d9a206e61a.png)